src/miral/application_selector.cpp

```diff
@@ -21,243 +21,439 @@
 
 using namespace miral;
 
-ApplicationSelector::ApplicationSelector(miral::WindowManagerTools const& in_tools) :
-    tools{in_tools}
-{}
+namespace miral
+{
+
+template <typename T>
+class WindowCyclingBehavior
+{
+public:
+    WindowCyclingBehavior(
+        std::function<bool(T const&, T const&)> comparison_func,
+        std::function<bool(T const&)> can_select,
+        std::function<void(T const&, T const&, T const&)> on_raised,
+        std::function<void(T const&)> on_cancelled,
+        std::function<void(T&)> reset
```

Comment:
OK, so this is what you were asking about? Declaring these as virtual functions vs providing them as functors?

I think what you actually want here is a traits class:

```c++
template<typename T> WindowCyclingTraits {};

struct WindowCyclingTraits<WeakApplication>
{
    static auto equals(WeakApplication const&, WeakApplication const&) => bool;
    static auto selectable(WeakApplication const&) => bool;
    ...
};
```

Then `WindowCyclingBehavior` can `using Traits = WindowCyclingTraits<T>` and `Traits::equals(...)`.

Reply:
That makes sense to me, but do the traits need to be `static`? The problem being that functions like `can_select` depend on `WindowManagerTools`, which is passed via a constructor. Would a non-static class defeat the purpose of the traits pattern?

Reply:
TPM traits are normally types or static functions. And if you're passing in `WindowManagerTools` to construct a "trait" instance, you might just as well pass it to the static method.

OTOH another design option is to template on a base class (which provides these functions), not the stored type.

Reply:
Also, things line `equals`/`comparison_func` could just be found by overloading.